update makefile and optimize assembly

mateoconlechuga opened the pull request:
The makefile changed to allow installation of the toolchain to be easier for users, plus also optimize the assembly routines to help with speed :)

mateoconlechuga committed:
update makefile for latest toolchain

the makefile changed to allow installation of the toolchain to be easier for users. this is documented in the release notes.

TheScienceElf commented:
Hey! Appreciate the help.

While the makefile does need to be updated for version v9.2, the issue I was describing is that the program gets caught in a loop in _fp_mul. Most likely I'm just violating calling conventions, but I haven't really taken the time to debug. At any rate, in the interest of having code that can be built with v9.1 to produce the correct result, I'm going to hold off on this commit until the bug gets fixed.

mateoconlechuga commented:
Oh that's odd! I can take a look later too see what might be the cause. Nice work on this, it is pretty neat :)

jacobly0 committed:
Fix asm structure.

calc84maniac committed:
Optimize asm math

mateoconlechuga commented:
Assembly optimization by @calc84maniac has program runtime down to ~547 seconds :)

EDIT: also the program works with the lastet master toolchain branch, so hopefully soon a new toolchain release will be available.

jacobly0 committed:
Use relative labels in asm.

jacobly0 committed:
Remove unnecessary cast.

jacobly0 committed:
Add size-optimized fp_sqrt which saves 67 bytes and is 12.5x faster than sqrtf.

mateoconlechuga commented:
Now @jacobly0's optimized sqrt has runtime down to ~512 seconds \o/

drdnar commented:
Here's some quick tips for eZ80 optimization. The optimized routines provided above supersede these, but they're good to help get you started.
* **IMPORTANT:** You *cannot* write to the stack under SP. The stack grows down and the interrupt handler *will* clobber those addresses. To use stack local variables, you must first adjust SP (either through pushing or something like `LD SP, IX`).
* `JR` is always slightly faster than `JP`. `JR` isn't always possible, but the assembler will convert it to `JP` if necessary.
* `XOR A, A` is slightly faster than `LD A, 0`, but it also resets the carry flag. (Even modern 64-bit compilers do this trick.)
* eZ80 does not require `$` for literals; e.g. on line 13 of asmmath.asm, you could write `LD IX, -6`. Without a prefix, numbers are interpreted as decimal, not hexadecimal.
* `LD reg8, reg8` is faster than `LD reg8, imm8`, so line 23 could be `LD DE, 0`, and then lines 31, 35, 39, 49, 53, and 57 would be `LD A, D` to make it slightly faster.
* IY and BC are not required to be preserved by the eZ80 calling convention. (Some OS routines do require IY to be set to `flags`, but the interrupt handler does not.)
* The `(IX/IY + offset)` addressing mode takes a *signed* 8-bit offset, so you can also do `(IX/IY - offset)`.
* A prolog that requires stack locals might look like `PUSH IX \ LD IX, -size_of_locals \ ADD IX, SP \ LD SP, IX \ LEA IX, IX+size_of_locals` and an epilog of `LD SP, IX \ POP IX \ RET`. Arguments can be accessed at positive offsets and locals at negative offsets.
* The eZ80 calling convention allows the callee function to clobber its arguments.
* [calc84maniac](https://github.com/calc84maniac)'s optimized routine uses IY as the stack frame pointer to avoid having to save and restore IX, which he can do because it's a leaf function; it doesn't call any other functions, so IY won't get clobbered.

jacobly0 committed:
Add size-optimized fp_div which saves 268 bytes and is almost 4x faster than using __ldivs and fp_mul.

jacobly0 committed:
Update comments.

jacobly0 committed:
Optimize fp_div to save 4 bytes and on average 100 cycles.

jacobly0 committed:
Optimize _fp_div to save 18 bytes and 18 cycles.

Could be faster, but the remaining speed opts
are no longer affecting render time significantly:

- unroll loops

- pre-round:

    @@ -249,5 +248,9 @@ _fp_div:
       ld l, a
    -  ld a, 24
    +  sbc hl, bc
    +  jq c, .no_carry
    +  inc de
    +.no_carry:
    +  ld a, 23
    +  add hl, hl
     .loop:
    -  adc hl, hl
       ex de, hl
    @@ -257,11 +260,9 @@ _fp_div:
       ex de, hl
    +  adc hl, hl
       dec a
       jr nz, .loop
    -  jr nc, .no_round
    -  inc hl
    -.no_round:
       ld a, (iy + 5)

mateoconlechuga committed:
update readme to allow for using the latest toolchain

Signed-off-by: mateoconlechuga <matthewwaltzis@gmail.com>

mateoconlechuga commented:
@TheScienceElf There is a [new toolchain release](https://github.com/CE-Programming/toolchain/releases/latest) available that fixes the building issue, so this patch is ready for a merge whenever. :) Hopefully the new toolchain and functions help improve the speed even more. Thanks for the awesome video!

TheScienceElf commented:
Hi Matt,

Thanks for the update! Unfortunately it looks like I'm having an issue with the linker when trying to build using the latest toolchain.

```
[linking] bin/RAYTRACE.bin
.../CEdev/meta/ld.alm [1493] macro ? [1244] :read? [7] obj/lto.src [264]:
        call _fp_sqr
call? [34] @ez80.word [9]
Processed: emit? size: value
Error: variable term used where not expected.
make: *** [.../CEdev/meta/makefile.mk:263: bin/RAYTRACE.bin] Error 2
```
Thanks to everyone for the optimizations/bugfixes! I'm eager to get this improved version out, but I'd like to make sure I can compile it locally first.

mateoconlechuga commented:
Hello! Yes, that is because the current assembly files in the repo never specified a section to link against, so with updated improvements to the linker it no longer just defaults to throwing everything in `.text` and hoping for the best - a section must be specified. I have added a [FAQ Entry](https://ce-programming.github.io/toolchain/v10.1/static/faq.html#my-code-used-to-compile) for this, and @jacobly0 added improved assembly docs so they can be always correct in the future. This PR already fixed that issue as one of the initial things.

As for this PR, you can check it out locally by following [this guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally), and then building using the [latest toolchain release](https://github.com/CE-Programming/toolchain/releases/tag/v10.1) - Note that v10.1 was just released as a bugfix over 10.0, so it is recommend to download/use that version. Thanks! :+1:

TheMiningTeamYT commented:
I can confirm the fixed point ASM from this pull request works with the latest version of the toolchain.
